File: 00/c.java
import java.util.*;

public class CacheMapping {
    static int cacheSize;
    static int blockSize;
    static int mappingType;
    static int replacementType;
    static Map<Integer, Block> cache;
    static int blockCounter = 0;
    static int ageCounter = 0;
    static Deque<Integer> FIFO_deque = new LinkedList<>();
    static Map<Integer, Integer> LRU_map = new HashMap<>();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the cache size: ");
        cacheSize = sc.nextInt();
        System.out.println("Enter the block size: ");
        blockSize = sc.nextInt();
        System.out.println("Select the mapping technique: ");
        System.out.println("1. Direct mapping");
        System.out.println("2. Set associative mapping");
        mappingType = sc.nextInt();
        if (mappingType == 2) {
            System.out.println("Select the replacement technique: ");
            System.out.println("1. FIFO");
            System.out.println("2. LRU");
            replacementType = sc.nextInt();
        }
        cache = new HashMap<Integer, Block>();
        while (true) {
            System.out.println("Enter a word to request: ");
            int word = sc.nextInt();
            int blockNum = getBlockNum(word);
            if (cache.containsKey(blockNum)) {
                System.out.println("Cache hit! Word found in block " + blockNum);
                if (mappingType == 2) {
                    if (replacementType == 1) {
                        updateFIFO(blockNum);
                    } else if (replacementType == 2) {
                        updateLRU(blockNum);
                    }
                }
            } else {
                System.out.println("Cache miss! Word not found in cache.");
                if (cache.size() == cacheSize) {
                    if (mappingType == 1) {
                        int oldestBlock = getOldestBlock();
                        cache.remove(oldestBlock);
                    } else if (mappingType == 2) {
                        int replacedBlock = getReplacedBlock();
                        cache.remove(replacedBlock);
                    }
                }
                Block newBlock = getBlockFromMemory(blockNum);
                cache.put(blockNum, newBlock);
                if (mappingType == 2) {
                    if (replacementType == 1) {
                        updateFIFO(blockNum);
                    } else if (replacementType == 2) {
                        updateLRU(blockNum);
                    }
                }
            }
            System.out.println("Cache content: " + cache.toString());
        }
    }

    public static int getBlockNum(int word) {
        return word / blockSize;
    }

    public static Block getBlockFromMemory(int blockNum) {
        // This is where you would retrieve the block from main memory
        return new Block(blockNum);
    }

    public static int getOldestBlock() {
        int oldestBlock =
cache.entrySet().iterator().next().getKey();
for (Map.Entry<Integer, Block> entry : cache.entrySet()) {
if (entry.getValue().getAge() > cache.get(oldestBlock).getAge()) {
oldestBlock = entry.getKey();
}
}
return oldestBlock;
}
public static int getReplacedBlock() {
    int replacedBlock = 0;
    if (replacementType == 1) {
        replacedBlock = FIFO_deque.pop();
        FIFO_deque.add(replacedBlock);
    } else if (replacementType == 2) {
        replacedBlock = getLRU();
    }
    return replacedBlock;
}

public static int getLRU() {
    int lruBlock = 0;
    int min = Integer.MAX_VALUE;
    for (Map.Entry<Integer, Integer> entry : LRU_map.entrySet()) {
        if (entry.getValue() < min) {
            min = entry.getValue();
            lruBlock = entry.getKey();
        }
    }
    return lruBlock;
}

public static void updateFIFO(int blockNum) {
    FIFO_deque.add(blockNum);
}

public static void updateLRU(int blockNum) {
    if(LRU_map.containsKey(blockNum)) {
        LRU_map.put(blockNum, ageCounter++);
    }else {
        LRU_map.put(blockNum, ageCounter++);
    }
}
}

class Block {
int blockNum;
int age;
public Block(int blockNum) {
    this.blockNum = blockNum;
    this.age = CacheMapping.blockCounter++;
}

public int getAge() {
    return this.age;
}

public String toString() {
    return "Block " + blockNum;
}
}

File: 00/c.py
import math

class CacheMapping:
    cacheSize = 0
    blockSize = 0
    mappingType = 0
    replacementType = 0
    cache = None
    blockCounter = 0
    ageCounter = 0
    FIFO_deque = []
    LRU_map = {}

    @staticmethod
    def main(args):
        
        print("Enter the cache size: ")
        CacheMapping.cacheSize = input()
        print("Enter the block size: ")
        CacheMapping.blockSize = input()
        print("Select the mapping technique: ")
        print("1. Direct mapping")
        print("2. Set associative mapping")
        CacheMapping.mappingType = input()
        if CacheMapping.mappingType == 2:
            print("Select the replacement technique: ")
            print("1. FIFO")
            print("2. LRU")
            CacheMapping.replacementType = input()
        CacheMapping.cache = {}
        while True:
            print("Enter a word to request: ")
            word = input()
            blockNum = CacheMapping.getBlockNum(word)
            if blockNum in CacheMapping.cache.keys():
                print("Cache hit! Word found in block " + str(blockNum))
                if CacheMapping.mappingType == 2:
                    if CacheMapping.replacementType == 1:
                        CacheMapping.updateFIFO(blockNum)
                    elif CacheMapping.replacementType == 2:
                        CacheMapping.updateLRU(blockNum)
            else:
                print("Cache miss! Word not found in cache.")
                if len(CacheMapping.cache) == CacheMapping.cacheSize:
                    if CacheMapping.mappingType == 1:
                        oldestBlock = CacheMapping.getOldestBlock()
                        CacheMapping.cache.pop(oldestBlock)
                    elif CacheMapping.mappingType == 2:
                        replacedBlock = CacheMapping.getReplacedBlock()
                        CacheMapping.cache.pop(replacedBlock)
                newBlock = CacheMapping.getBlockFromMemory(blockNum)
                CacheMapping.cache[blockNum] = newBlock
                if CacheMapping.mappingType == 2:
                    if CacheMapping.replacementType == 1:
                        CacheMapping.updateFIFO(blockNum)
                    elif CacheMapping.replacementType == 2:
                        CacheMapping.updateLRU(blockNum)
            print("Cache content: " + str(CacheMapping.cache))

    @staticmethod
    def getBlockNum(word):
        return math.trunc(word / float(CacheMapping.blockSize))

    @staticmethod
    def getBlockFromMemory(blockNum):
        # This is where you would retrieve the block from main memory
        return Block(blockNum)

    @staticmethod
    def getOldestBlock():
        oldestBlock = CacheMapping.cache.entrySet().iterator().next().getKey()
        for entry in CacheMapping.cache.entrySet():
            if entry.getValue().getAge() > CacheMapping.cache[oldestBlock].getAge():
                oldestBlock = entry.getKey()
        return oldestBlock
    @staticmethod
    def getReplacedBlock():
        replacedBlock = 0
        if CacheMapping.replacementType == 1:
            replacedBlock = CacheMapping.FIFO_deque.pop()
            CacheMapping.FIFO_deque.add(replacedBlock)
        elif CacheMapping.replacementType == 2:
            replacedBlock = CacheMapping.getLRU()
        return replacedBlock

    @staticmethod
    def getLRU():
        lruBlock = 0
        min = Integer.MAX_VALUE
        for entry in CacheMapping.LRU_map.entrySet():
            if entry.getValue() < min:
                min = entry.getValue()
                lruBlock = entry.getKey()
        return lruBlock

    @staticmethod
    def updateFIFO(blockNum):
        CacheMapping.FIFO_deque.add(blockNum)

    @staticmethod
    def updateLRU(blockNum):
        if blockNum in CacheMapping.LRU_map.keys():

            CacheMapping.LRU_map[blockNum] = CacheMapping.ageCounter
            CacheMapping.ageCounter += 1
        else:
            CacheMapping.LRU_map[blockNum] = CacheMapping.ageCounter
            CacheMapping.ageCounter += 1

class Block(object) :
    blockNum = 0
    age = 0
    def __init__(self, blockNum) :
        self.blockNum = blockNum
        self.age = CacheMapping.blockCounter +12 1
    def  getAge(self) :
        return self.age
    def  toString(self) :
        return "Block " + str(self.blockNum)        

def main():
    CacheMapping.main([])

if __name__ == "__main__":
    main()
